compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,857 @@
+  std::string gen;
+
+  if (this->debugPrintlines) {
+    gen += this->genDefine("CHPL_MLI_DEBUG_PRINT");
+  }
+
+  gen += this->genHeaderInc("stdlib.h", true);
+  gen += this->genHeaderInc("stdio.h", true);
+  gen += this->genHeaderInc("zmq.h", true);
+  gen += this->genHeaderInc("chpl__header.h");
+  gen += this->genNote("We use Makefile magic to make this visible!");
+  gen += this->genHeaderInc("mli_common_code.c");
+  gen += "\n";
+
+  for (i = this->typeMap.begin(); i != this->typeMap.end(); ++i) {
+    if (this->debugPrintlines) {
+      std::string tpn = this->genTypeName(i->first);
+      gen += this->genComment(tpn.c_str());
+    }
+
+    gen += this->genMarshalPushRoutine(i->first);
+    gen += this->genMarshalPullRoutine(i->first);
+  }
+
+  this->setOutputAndWrite(&this->fiMarshalling, gen);
+
+  return;
+}
+
+std::string MLIContext::genMarshalBodyPrimitiveScalar(Type* t, bool out) {
+  std::string gen;
+
+  // On pack, target buffer is input parameter. On unpack, the temporary.
+  const char* target = out ? "obj" : "result";
+
+  // Move the raw bytes of the type to/from the wire.
+  gen += "skt_err=";
+  gen += this->genSocketCall("skt", target, out);
+
+  // Generate a null frame in the opposite direction for the ACK.
+  gen += this->genSocketCall("skt", NULL, not out);
+
+  return gen;
+}
+
+std::string MLIContext::genMarshalRoutine(Type* t, bool out) {
+  int64_t id = this->assignUniqueTypeID(t);
+  std::string gen;
+
+  // Push returns nothing, while pull returns the type being read in.
+  if (out) {
+    gen += "void ";
+  } else {
+    gen += this->genTypeName(t);
+    gen += " ";
+  }
+
+  // Select appropriate prefix for function name based on direction.
+  gen += out ? marshal_push_prefix : marshal_pull_prefix;
+  gen += str(id);
+  gen += "(void* skt";
+
+  // Push routines expect the type as a parameter (named "obj").
+  if (out) {
+    gen += ",";
+    gen += this->genTypeName(t);
+    gen += " obj";
+  }
+
+  gen += ")";
+  gen += scope_begin;
+
+  // Always declare a variable to catch socket errors.
+  gen += this->genNewDecl("int", "skt_err");
+
+  // If unpacking, declare a temporary for the return value.
+  if (not out) {
+    gen += this->genNewDecl(t, "result");
+  }
+
+  // If allocating, declare a temporary for amount.
+  if (this->typeRequiresAllocation(t)) {
+    gen += this->genNewDecl("uint64_t", "bytes");
+  }
+
+  // Insert a debug message if appropriate.
+  if (this->debugPrintlines) {
+    std::string msg;
+
+    msg += out ? "Pushing type: " : "Pulling type: ";
+    msg += this->genTypeName(t);
+
+    gen += this->genDebugPrintCall(msg.c_str(), "[RPC]");
+  }
+
+  //
+  // Handle translation of different type classes here. Note that right now
+  // the only things we can translate are primitive scalars.
+  //
+  if (isPrimitiveScalar(t)) {
+    gen += this->genMarshalBodyPrimitiveScalar(t, out);
+  } else {
+    USR_FATAL("MLI does not support code generation for type", t);
```

Comment:
For a user facing error messages, please don't use acronyms which haven't been explained in the documentation.

Reply:
Hopefully I've addressed all of these by now!